File: src/number_square_py/number_square_py/number_publisher.py
import rclpy  #Importing the ros client library for python
from rclpy.node import Node #Importing the Node class for creating ROS2 Nodes
from std_msgs.msg import Int32 #Importing the standard message type 32-bit integer
#Importing required dependencies for a QoS profile
from rclpy.qos import QoSProfile, ReliabilityPolicy, DurabilityPolicy

#Creating a node number publisher
class NumberPublisher(Node):

    def __init__(self):
        #Creates a node names number_publisher
        super().__init__('number_publisher')
        qos_profile = QoSProfile(
            reliability = ReliabilityPolicy.RELIABLE,
            durability = DurabilityPolicy.VOLATILE, 
            depth = 10
        )

        """
        Creating a publisher of 
        Topic : number
        Message data type: Int32
        Queue Size : 10"""
        self.publisher_ = self.create_publisher(Int32, 'number', qos_profile)

        self.get_logger().info("Starting automatic number publishing.")

        self.count = 0  # Initialize the counter

        # Create a timer to publish number every 1 second (1000 milliseconds)
        self.timer = self.create_timer(1.0, self.timer_callback)

    def timer_callback(self):
        #Publish numbers in increasing order automatically
        msg = Int32()
        msg.data = self.count
        self.publisher_.publish(msg)
        self.get_logger().info(f"Publishing: {msg.data}")
        self.count += 1  # Increment the counter

def main(args=None):
    #Starting ROS2 communication
    rclpy.init(args=args)
    #Initialize the node
    number_publisher = NumberPublisher()
    #Spin keeps the node alive and processing timers/callbacks
    rclpy.spin(number_publisher)

    #Clean up resources once node is closed
    number_publisher.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: src/number_square_cpp/src/number_publisher.cpp
#include "rclcpp/rclcpp.hpp" //Importing ros2 c++ client library
#include "std_msgs/msg/int32.hpp" //Importing ros2 standard integer message type
#include <iostream> //Importing c++ input/output library

using namespace std::chrono_literals;

//Creating NumberPublisher class inheriting from Node class
class NumberPublisher : public rclcpp::Node {
public:
    //Constructing the Node with name "name_publisher"
    NumberPublisher() : Node("number_publisher"), count_(0) {

        auto qos = rclcpp::QoS(rclcpp::QoSInitialization::from_rmw(rmw_qos_profile_default))
        .reliable()
        .durability_volatile()
        .keep_last(10);

        //Creating a publisher that publishes to topic 'number using Int32 and has a queue size of 10
        publisher_ = this -> create_publisher<std_msgs::msg::Int32>("number", qos);
        //Creates a timer that triggers every second
        timer_ = this->create_wall_timer(
            1000ms, // 1 second interval
            std::bind(&NumberPublisher::timer_callback, this));
        
        RCLCPP_INFO(this->get_logger(), "Number publisher started with timer...");
    }

private:
    void timer_callback() {
        //Creating a message and assigning count_ to it
        auto message = std_msgs::msg::Int32();
        message.data = count_++;
        
        // Logging and Publishing the number
        RCLCPP_INFO(this->get_logger(), "Published: %d", message.data);
        publisher_->publish(message);
    }
    
    int count_;
    rclcpp::Publisher<std_msgs::msg::Int32>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};


int main(int argc, char *argv[]) {
    //Initializes ros2
    rclcpp::init(argc, argv);
    //Creating node instance
    rclcpp::spin(std::make_shared<NumberPublisher>());
    //Closing the node
    rclcpp::shutdown();
    return 0;
}
